monai/data/dataset_calculator.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+import multiprocessing as mp
+from typing import Dict, Sequence
+
+import numpy as np
+
+from monai.transforms import LoadImaged
+
+
+class DatasetCalculator:
+    """
+    This class provides a way to calculate a reasonable output voxel spacing according to
+    the input dataset. The achieved values can used to resample the input in 3d segmentation tasks
+    (like using as the `pixel` parameter in `monai.transforms.Spacingd`).
+    In addition, it also supports to count the mean, std, min and max intensities of the input,
+    and these statistics are helpful for image normalization
+    (like using in `monai.transforms.ScaleIntensityRanged` and `monai.transforms.NormalizeIntensityd`).
+
+    The algorithm for calculation refers to:
+    `Automated Design of Deep Learning Methods for Biomedical Image Segmentation <https://arxiv.org/abs/1904.08128>`_.
+
+    """
+
+    def __init__(
+        self,
+        datalist: Sequence[Dict],
+        image_key: str = "image",
+        label_key: str = "label",
+        meta_key_postfix: str = "meta_dict",
+        num_processes: int = 1,
+    ):
+        """
+        Args:
+            datalist: a list that contains the path of all images and labels. The list is
+                consisted with dictionaries, and each dictionary contains the image and label
+                path of one sample. For datasets that have Decathlon format, datalist can be
+                achieved by calling `monai.data.load_decathlon_datalist`.
+            image_key: the key name of images. Defaults to `image`.
+            label_key: the key name of labels. Defaults to `label`.
+            meta_key_postfix: for nifti images, use `{image_key}_{meta_key_postfix}` to
+                store the metadata of images.
+            num_workers: the maximum number of processes can be used in data loading.
+
+        """
+
+        self.datalist = datalist
+        self.image_key = image_key
+        self.label_key = label_key
+        self.meta_key_postfix = meta_key_postfix
+        self.num_processes = num_processes
+        self.loader = LoadImaged(keys=[image_key, label_key], meta_key_postfix=meta_key_postfix)
```

Comment:
I'd suggest `loader` should be an argument so that we don't hard code using `LoadImaged`. This would allow us to use input datasets that aren't just dictionaries for one.

Reply:
Fixed.